python/cudf/cudf/core/buffer/utils.py

```diff
@@ -0,0 +1,218 @@
+    allocation. Otherwise, the data of `data` is **not** copied, instead the
+    new buffer keeps a reference to `data` in order to retain the lifetime
+    of `data`.
+
+    If `data` is an integer it must represent a device pointer and `size` must
+    be specified. If `data` isn't an integer both `size` and `owner` must be
+    None.
+
+    Raises ValueError if the data of `data` isn't C-contiguous.
+
+    Parameters
+    ----------
+    data : int or buffer-like or array-like
+        - An integer representing a pointer to device memory, or
+        - An object that exposes either device or host memory through
+          `__array_interface__`, `__cuda_array_interface__`, or the buffer
+          protocol.
+        If `data` represents host memory, data will be copied to device.
+    exposed : bool, optional
+        Whether or not a raw pointer (integer or C pointer) has
+        been exposed to the outside world. If this is the case,
+        the buffer cannot be spilled.
+    size : int, optional
+        Size of memory in bytes. Must be specified if `data` is an integer
+        otherwise it must be None.
+    owner : object, optional
+        Python object to which the lifetime of the memory allocation is tied.
+        A reference to this object is kept in the returned Buffer. Can only be
+        specified when `data` is an integer otherwise the `data` itself will be
+        set as the owner.
+
+    Return
+    ------
+    DeviceBufferLike
+        A device-buffer-like instance that represents the device memory
+        of `data`.
+    """
+
+    if isinstance(data, DeviceBufferLike):
+        return data
+
+    data = ensure_buffer_like(data=data, size=size, owner=owner)
+    manager = global_manager_get()
+    if manager is None:
+        return Buffer(data)
+    return SpillableBuffer(data=data, exposed=exposed, manager=manager)
+
+
+def get_columns(obj: Any) -> List[Column]:
+    from cudf._lib.column import Column
+    from cudf.core.column_accessor import ColumnAccessor
+    from cudf.core.frame import Frame
+    from cudf.core.indexed_frame import IndexedFrame
+
+    """Return all columns in `obj` (no duplicates)"""
+    found: List[Column] = []
+    found_ids: Set[int] = set()
+
+    def _get_columns(obj: object) -> None:
+        if isinstance(obj, Column):
+            if id(obj) not in found_ids:
+                found_ids.add(id(obj))
+                found.append(obj)
+        elif isinstance(obj, IndexedFrame):
+            _get_columns(obj._data)
+            _get_columns(obj._index)
+        elif isinstance(obj, Frame):
+            _get_columns(obj._data)
+        elif isinstance(obj, ColumnAccessor):
+            for o in obj.columns:
+                _get_columns(o)
+        elif isinstance(obj, (list, tuple)):
+            for o in obj:
+                _get_columns(o)
+        elif isinstance(obj, Mapping):
+            for o in obj.values():
+                _get_columns(o)
```

Comment:
Do we really need to handle all of these cases? Our usage of `with_spill_lock` should be restricted to a pretty small set of possible inputs since it's at the Cython level, right? Most of these branches probably indicate to me that an API is still using some old calling convention.